### PR TITLE
Define unknownFieldRetention trait

### PR DESCRIPTION
**Pull request by dhpiggott**

### CODE REVIEW

File: modules/core/resources/META-INF/smithy/unknownfieldretention.smithy [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    selector: "structure > member :test(> document)"
+)
+structure unknownDocumentFieldRetention {}
+
+/// Retain unknown fields of a containing structure in a map.
+/// This variant controls the behaviour for JSON codecs.
+@trait(
+    selector: "structure > member :test(> document)"
+)
+structure unknownJsonFieldRetention {}
```
Comment (dhpiggott) [inline modules/core/resources/META-INF/smithy/unknownfieldretention.smithy:17]: From @Baccata via Slack:
> just unify the two traits as jsonUnkown. Document abides by json semantics, and we do implement the same semantics in both Document and jsoniter

File: modules/core/test/src/alloy/UnknownFieldRetentionTraitProviderSpec.scala [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+          .builder()
+          .id(targetId)
+          .target(documentShape.getId)
+          .addTrait(new UnknownFieldRetentionTrait)
+          .build()
+      )
+      .build()
+
+    val model =
+      Model.assembler.disableValidation
```
[the remark this answers is not in the document]
Comment (dhpiggott) [reply modules/core/test/src/alloy/UnknownFieldRetentionTraitProviderSpec.scala]: Not sure, I based this spec on one of the existing ones which had this. I'll see what happens if I enable it.

**Comment by dhpiggott** [comment]
@lewisjkl GitHub told me you merged this but it also tells me that https://github.com/disneystreaming/alloy/pull/180 replaced it. So wondering if you meant to merge this and if so how the two relate. I see:
> [lewisjkl](https://github.com/lewisjkl) merged commit [e270be0](https://github.com/disneystreaming/alloy/commit/e270be09258b10cad0091e0c39c92d29f3bcf03b) into main  [14 hours ago](https://github.com/disneystreaming/alloy/pull/135#event-13867096213)
But then I see nothing on main 🤔 Bit confused!

**Comment by Baccata** [comment]
Ben's PR forked from your branch, and had all the commits your branched contained. So merging Ben's PR into main implies  merging this PR into main, as all the commits end up in main.

**Comment by dhpiggott** [comment]
Aaaah, thanks! I should have looked at https://github.com/disneystreaming/alloy/pull/180 more carefully, then I'd have realised.

**Pull request merged**